File: locationpub/locationpub/locationpub_node.py
import math

import rclpy
from geometry_msgs.msg import TransformStamped
from rclpy.node import Node
from tf2_ros import TransformBroadcaster


def quaternion_from_yaw(yaw):
    half_yaw = yaw * 0.5
    return 0.0, 0.0, math.sin(half_yaw), math.cos(half_yaw)


class LocationPub(Node):
    def __init__(self):
        super().__init__('locationpub')

        self.declare_parameter('parent_frame', 'map')
        self.declare_parameter('child_frame', 'turtlebot3_waffle')
        self.declare_parameter('x', 0.569)
        self.declare_parameter('y', 0.541)
        self.declare_parameter('z', 0.0)
        self.declare_parameter('yaw', 0.0)
        self.declare_parameter('publish_rate', 30.0)

        self.parent_frame = self.get_parameter('parent_frame').value
        self.child_frame = self.get_parameter('child_frame').value
        self.x = float(self.get_parameter('x').value)
        self.y = float(self.get_parameter('y').value)
        self.z = float(self.get_parameter('z').value)
        self.yaw = float(self.get_parameter('yaw').value)
        publish_rate = float(self.get_parameter('publish_rate').value)

        if publish_rate <= 0.0:
            raise ValueError('publish_rate must be greater than 0.0')

        self.tf_broadcaster = TransformBroadcaster(self)
        self.timer = self.create_timer(1.0 / publish_rate, self.publish_tf)

        self.get_logger().info(
            f'Publishing localization TF {self.parent_frame} -> {self.child_frame} '
            f'at {self.x:.3f}, {self.y:.3f}, {self.z:.3f} yaw {self.yaw:.3f}')

    def publish_tf(self):
        transform = TransformStamped()
        transform.header.stamp = self.get_clock().now().to_msg()
        transform.header.frame_id = self.parent_frame
        transform.child_frame_id = self.child_frame
        transform.transform.translation.x = self.x
        transform.transform.translation.y = self.y
        transform.transform.translation.z = self.z

        qx, qy, qz, qw = quaternion_from_yaw(self.yaw)
        transform.transform.rotation.x = qx
        transform.transform.rotation.y = qy
        transform.transform.rotation.z = qz
        transform.transform.rotation.w = qw

        self.tf_broadcaster.sendTransform(transform)


def main(args=None):
    rclpy.init(args=args)
    node = LocationPub()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: locationpub/src/locationpub_node.cpp
#include <chrono>
#include <cmath>
#include <memory>
#include <stdexcept>
#include <string>

#include "geometry_msgs/msg/transform_stamped.hpp"
#include "rclcpp/rclcpp.hpp"
#include "tf2_ros/transform_broadcaster.h"

using namespace std::chrono_literals;

namespace
{

geometry_msgs::msg::Quaternion quaternionFromYaw(const double yaw)
{
  geometry_msgs::msg::Quaternion quaternion;
  const double half_yaw = yaw * 0.5;
  quaternion.x = 0.0;
  quaternion.y = 0.0;
  quaternion.z = std::sin(half_yaw);
  quaternion.w = std::cos(half_yaw);
  return quaternion;
}

}  // namespace

class LocationPub : public rclcpp::Node
{
public:
  LocationPub()
  : Node("locationpub")
  {
    parent_frame_ = declare_parameter<std::string>("parent_frame", "map");
    child_frame_ = declare_parameter<std::string>("child_frame", "turtlebot3_waffle");
    x_ = declare_parameter<double>("x", 0.569);
    y_ = declare_parameter<double>("y", 0.541);
    z_ = declare_parameter<double>("z", 0.0);
    yaw_ = declare_parameter<double>("yaw", 0.0);
    tf_time_offset_ = declare_parameter<double>("tf_time_offset", 0.2);
    const double publish_rate = declare_parameter<double>("publish_rate", 30.0);

    if (publish_rate <= 0.0) {
      throw std::invalid_argument("publish_rate must be greater than 0.0");
    }
    if (tf_time_offset_ < 0.0) {
      throw std::invalid_argument("tf_time_offset must be greater than or equal to 0.0");
    }

    tf_broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);
    const auto period = std::chrono::duration<double>(1.0 / publish_rate);
    timer_ = create_wall_timer(
      std::chrono::duration_cast<std::chrono::nanoseconds>(period),
      std::bind(&LocationPub::publishTf, this));

    RCLCPP_INFO(
      get_logger(),
      "Publishing localization TF %s -> %s at %.3f, %.3f, %.3f yaw %.3f with %.3fs stamp offset",
      parent_frame_.c_str(), child_frame_.c_str(), x_, y_, z_, yaw_, tf_time_offset_);
  }

private:
  void publishTf()
  {
    geometry_msgs::msg::TransformStamped transform;
    const auto stamp_offset = std::chrono::duration<double>(tf_time_offset_);
    const auto stamp = now() - rclcpp::Duration(
      std::chrono::duration_cast<std::chrono::nanoseconds>(stamp_offset));
    const auto stamp_ns = stamp.nanoseconds();
    transform.header.stamp.sec = static_cast<int32_t>(stamp_ns / 1000000000);
    transform.header.stamp.nanosec = static_cast<uint32_t>(stamp_ns % 1000000000);
    transform.header.frame_id = parent_frame_;
    transform.child_frame_id = child_frame_;
    transform.transform.translation.x = x_;
    transform.transform.translation.y = y_;
    transform.transform.translation.z = z_;
    transform.transform.rotation = quaternionFromYaw(yaw_);

    tf_broadcaster_->sendTransform(transform);
  }

  std::string parent_frame_;
  std::string child_frame_;
  double x_;
  double y_;
  double z_;
  double yaw_;
  double tf_time_offset_;
  std::unique_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<LocationPub>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
